File: flink-app/src/main/java/com/example/ctr/domain/model/CTRResult.java
package com.example.ctr.domain.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CTRResult {
    private String productId;
    private long impressions;
    private long clicks;
    private double ctr;
    private long windowStart;
    private long windowEnd;

    public static CTRResult calculate(String productId, long impressions, long clicks, long windowStart,
            long windowEnd) {
        double ctr = (impressions == 0) ? 0.0 : (double) clicks / impressions;
        return CTRResult.builder()
                .productId(productId)
                .impressions(impressions)
                .clicks(clicks)
                .ctr(ctr)
                .windowStart(windowStart)
                .windowEnd(windowEnd)
                .build();
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/domain/model/CTRResult.kt
package com.example.ctr.domain.model

/**
 * CTR aggregation result for a particular product window.
 */
data class CTRResult(
    val productId: String,
    val impressions: Long,
    val clicks: Long,
    val ctr: Double,
    val windowStart: Long,
    val windowEnd: Long
) {

    companion object {
        fun calculate(productId: String, impressions: Long, clicks: Long, windowStart: Long, windowEnd: Long): CTRResult {
            val ctrValue = if (impressions == 0L) 0.0 else clicks.toDouble() / impressions
            return CTRResult(
                productId = productId,
                impressions = impressions,
                clicks = clicks,
                ctr = ctrValue,
                windowStart = windowStart,
                windowEnd = windowEnd
            )
        }
    }
}
